<launch>
    <!-- Task and logging arguments -->
    <arg name="task_type"/>
    <arg name="bandits_logging_enabled"                 default="true"/>
    <arg name="controller_logging_enabled"              default="false"/>
    <arg name="test_id"                                 default="default"/>
    <arg name="task_max_time_override"                  default="false"/>
    <arg name="task_max_time"                           if="$(arg task_max_time_override)"/>

    <!-- Simulator arguments -->
    <arg name="launch_simulator"                        default="true"/>
    <arg name="start_bullet_viewer"                     default="true"/>
    <arg name="screenshots_enabled"                     default="false"/>

    <!-- Planner trial type -->
    <arg name="trial_type"                      default="diminishing_rigidity_single_model_least_squares_stretching_constraint_controller"/>
    <!--<arg name="trial_type"                      default="diminishing_rigidity_single_model_least_squares_stretching_avoidance_controller"/>-->
    <!--<arg name="trial_type"                      default="constraint_single_model_constraint_controller"/>-->
    <!--<arg name="trial_type"                      default="multi_model_bandit_test"/>-->

    <!-- Planner visualization arguments -->
    <arg name="disable_all_visualizations"              default="false"/>
    <arg name="visualize_object_desired_motion"         default="false"/>
    <arg name="visualize_gripper_motion"                default="false"/>
    <arg name="visualize_object_predicted_motion"       default="false"/>
    <arg name="visualize_prm"                           default="false"/>
    <arg name="visualize_rrt"                           default="true"/>
    <arg name="visualize_free_space_graph"              default="false"/>
    <arg name="visualize_correspondences"               default="false"/>

    <!-- Model/Controller arguments -->
    <!-- Diminishing rigidity -->
    <arg name="deformability_override"                  default="false"/>
    <arg name="translational_deformability"             if="$(arg deformability_override)"/>
    <arg name="rotational_deformability"                if="$(arg deformability_override)"/>
    <!-- Adaptive Jacobian -->
    <arg name="adaptive_model_learning_rate"            default="0.000001"/>
    <!-- Directional rigidity -->
    <arg name="translational_dir_deformability"         default="4.0"/>
    <arg name="translational_dis_deformability"         default="10.0"/>
    <arg name="rotational_dis_deformability"            default="20.0"/>
    <!-- Stretching Avoidance Controller -->
    <arg name="stretching_cosine_threshold_override"    default="false"/>
    <arg name="stretching_cosine_threshold"             if="$(arg stretching_cosine_threshold_override)"/>
    <!-- Multi-model Bandit  -->
    <arg name="kalman_parameters_override"              default="false"/>
    <arg name="process_noise_factor"                    if="$(arg kalman_parameters_override)"/>
    <arg name="observation_noise_factor"                if="$(arg kalman_parameters_override)"/>
    <arg name="bandit_algorithm"                        default="UCB"/>
    <arg name="correlation_strength_factor_override"    default="false"/>
    <arg name="correlation_strength_factor"             if="$(arg correlation_strength_factor_override)"/>

    <!-- Planner internal arguments -->
    <arg name="launch_planner"                          default="true"/>
    <arg name="collect_results_for_all_models"          default="false"/>
    <arg name="use_random_seed"                         default="false"/>
    <arg name="static_seed_override"                    default="false"/>

    <!-- Triggers transition on smoothing on 2nd plan, not on 3rd -->
    <!--<arg name="static_seed"                             default="156f1ba9b0fbafe2" if="$(arg static_seed_override)"/>-->
    <arg name="static_seed"                             default="156f1ba9b0fbafe2" if="$(arg static_seed_override)"/>

    <!-- Setup task parameters -->
    <include file="$(find deformable_manipulation_experiment_params)/launch/$(arg task_type)_params.launch"/>
    <param name="log_folder"                            value="$(find smmap)/logs/$(arg task_type)/$(arg test_id)/"         type="string"/>

    <!-- Setup the simulator -->
    <group ns="deform_simulator_node">
        <param name="start_bullet_viewer"               value="$(arg start_bullet_viewer)"      type="bool"/>
        <param name="screenshots_enabled"               value="$(arg screenshots_enabled)"      type="bool"/>
        <param name="settle_time"                       value="4.0"                             type="double"/>
        <param name="tf_wait_time"                      value="0.0"                             type="double"/>
        <param name="display_grippers_axes"             value="false"                           type="bool"/>
    </group>
    <env name="OSG_FILE_PATH" value="$(find OpenSceneGraph)/data"/>
    <node name="deform_simulator_node" pkg="deform_simulator" type="custom_scene_node" required="true" if="$(arg launch_simulator)"/>


    <!-- Setup the planner -->
    <group ns="smmap_planner_node">
        <param name="trial_type"                        value="$(arg trial_type)"                           type="string"/>
        <param name="use_random_seed"                   value="$(arg use_random_seed)"                      type="bool"/>
        <param name="static_seed"                       value="$(arg static_seed)"                          type="string"   if="$(arg static_seed_override)"/>

        <param name="bandits_logging_enabled"           value="$(arg bandits_logging_enabled)"              type="bool"/>
        <param name="controller_logging_enabled"        value="$(arg controller_logging_enabled)"           type="bool"/>

        <param name="use_stored_world_state"                value="false"                                   type="bool"/>
        <param name="world_state_file_name_prefix"          value="world_state"                             type="string"/>
        <param name="world_state_file_name_suffix_to_load"  value="band_weirdness_controller_only"          type="string"/>

        <param name="disable_all_visualizations"        value="$(arg disable_all_visualizations)"           type="bool"/>
        <param name="visualize_object_desired_motion"   value="$(arg visualize_object_desired_motion)"      type="bool"/>
        <param name="visualize_gripper_motion"          value="$(arg visualize_gripper_motion)"             type="bool"/>
        <param name="visualize_object_predicted_motion" value="$(arg visualize_object_predicted_motion)"    type="bool"/>
        <param name="visualize_prm"                     value="$(arg visualize_prm)"                        type="bool"/>
        <param name="visualize_rrt"                     value="$(arg visualize_rrt)"                        type="bool"/>
        <param name="visualize_free_space_graph"        value="$(arg visualize_free_space_graph)"           type="bool"/>
        <param name="visualize_correspondences"         value="$(arg visualize_correspondences)"            type="bool"/>

        <group ns="rrt">
            <param name="timeout"                           value="6000000000.0"                   type="double"/>
            <param name="reuse_old_results"                 value="false"                   type="bool"/>
            <param name="store_new_results"                 value="false"                   type="bool"/>

            <param name="use_brute_force_nn"                value="false"                   type="bool"/>
            <param name="kd_tree_grow_threshold"            value="500"                     type="int"/>

            <param name="forward_tree_extend_iterations"    value="10"                      type="int"/>
            <param name="backward_tree_extend_iterations"   value="10"                      type="int"/>
            <param name="goal_bias"                         value="0.10"                    type="double"/>
            <param name="use_cbirrt_style_projection"       value="false"                   type="bool"/>
            <param name="best_near_radius"                  value="0.001"                   type="double"/>

            <param name="band_dist2_scaling_factor"         value="0.000001"                type="double"/>
            <param name="band_max_points_"                  value="500"                     type="double"/>

            <param name="feasibility_distance_scale_factor" value="0.0"                     type="double"/>

            <!-- Used to test shortcut smoothing -->
            <param name="use_stored_path"                   value="false"                   type="bool"/>
            <param name="path_file_name_prefix"             value="rrt_path"                type="string"/>
            <param name="path_file_name_suffix_to_load"     value="before_blacklisting_root_idx_4__2019-01-24__13-55-48"     type="string"/>
        </group>

        <group ns="band">
            <param name="logging_enabled"               value="$(arg bandits_logging_enabled)"                              type="bool"/>
            <param name="log_folder"                    value="$(find smmap)/logs/$(arg task_type)/$(arg test_id)/"         type="string"/>
            <param name="use_stored_band"               value="false"                                                       type="bool"/>
            <param name="band_file_name_prefix"         value="band"                                                        type="string"/>
            <param name="band_file_name_suffix_to_load" value="stuck_on_edge_of_obstacle"                                   type="string"/>
        </group>

        <group ns="transition_estimation">
            <param name="default_propagation_confidence"        value="1.0"                     type="double"/>
            <param name="default_band_dist_threshold"           value="0.2"                     type="double"/>
            <param name="confidence_threshold"                  value="0.2"                     type="double"/>
            <param name="template_misalignment_scale_factor"    value="1.0"                     type="double"/>
            <param name="tighten_delta_scale_factor"            value="1.0"                     type="double"/>
            <param name="homotopy_changes_scale_factor"         value="1.0"                     type="double"/>
            <param name="use_stored_transitions"                value="false"                   type="bool"/>
            <param name="file_name_prefix"                      value="../learned_transitions"  type="string"/>
            <rosparam>
                file_name_suffixes_to_load: ["thesis_proposal_example"]
            </rosparam>
        </group>

        <!-- Dinimishing Rigidity parameters and overrides -->
        <group ns="diminishing_rigidity">
            <param name="translational_deformability"   value="$(arg translational_deformability)"      type="double"   if="$(arg deformability_override)"/>
            <param name="rotational_deformability"      value="$(arg rotational_deformability)"         type="double"   if="$(arg deformability_override)"/>
        </group>

        <!-- Adaptive Jacobian parameters and overrides -->
        <group ns="adaptive_model">
            <param name="adaptive_model_learning_rate"  value="$(arg adaptive_model_learning_rate)"     type="double"/>
        </group>

        <!-- Constraint Model parameters and overrides -->
        <group ns="constraint_model">
            <param name="translational_dir_deformability"   value="$(arg translational_dir_deformability)"  type="double"/>
            <param name="translational_dis_deformability"   value="$(arg translational_dis_deformability)"  type="double"/>
            <param name="rotational_dis_deformability"      value="$(arg rotational_dis_deformability)"     type="double"/>
        </group>

        <!-- Multi-model parameters and overrides -->
        <group ns="multi_model">
            <param name="bandit_algorithm"                  value="$(arg bandit_algorithm)"                 type="string"/>
            <param name="process_noise_factor"              value="$(arg process_noise_factor)"             type="double"   if="$(arg kalman_parameters_override)"/>
            <param name="observation_noise_factor"          value="$(arg observation_noise_factor)"         type="double"   if="$(arg kalman_parameters_override)"/>
            <param name="collect_results_for_all_models"    value="$(arg collect_results_for_all_models)"   type="bool"/>
            <param name="correlation_strength_factor"       value="$(arg correlation_strength_factor)"      type="double"   if="$(arg correlation_strength_factor_override)"/>
            <param name="reward_scale_annealing_factor"     value="0.9"                                     type="double"/>
        </group>

        <!-- Old pure Jacobian based controller -->
        <group ns="jacobian_controller">
            <param name="optimization_enabled"          value="true"                                    type="bool"/>
        </group>

        <!-- Stretching avoidance controller arguments -->
        <group ns="stretching_avoidance_controller">
            <!--<param name="solver_type"                   value="nomad_optimization"                      type="string"/>-->
            <!--<param name="solver_type"                   value="random_sampling"                         type="string"/>-->
            <param name="solver_type"                   value="gradient_descent"                        type="string"/>
            <param name="max_sampling_counts"           value="5000"                                    type="int"/>
        </group>
    </group>
    <node name="smmap_planner_node" pkg="smmap" type="planner_test_node" required="true" if="$(arg launch_planner)"/>
</launch>
